kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/aggregation/settings/KoverSettingsGradlePlugin.kt

```diff
@@ -0,0 +1,117 @@
+import kotlinx.kover.gradle.aggregation.settings.tasks.KoverHtmlReportTask
+import kotlinx.kover.gradle.aggregation.settings.tasks.KoverXmlReportTask
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.initialization.ProjectDescriptor
+import org.gradle.api.initialization.Settings
+import org.gradle.api.model.ObjectFactory
+import org.gradle.kotlin.dsl.apply
+import org.gradle.kotlin.dsl.create
+import org.gradle.kotlin.dsl.register
+import org.gradle.kotlin.dsl.support.serviceOf
+
+internal class KoverSettingsGradlePlugin: Plugin<Settings> {
+
+    override fun apply(target: Settings) {
+        val objects = target.serviceOf<ObjectFactory>()
+
+        val settingsExtension = target.extensions.create<KoverSettingsExtensionImpl>(KoverNames.settingsExtensionName, objects)
+
+        target.gradle.settingsEvaluated {
+            KoverParametersProcessor.process(settingsExtension, providers)
+        }
+
+        target.gradle.beforeProject {
+            if (!settingsExtension.coverageIsEnabled.get()) {
+                return@beforeProject
+            }
+
+            val agentDependency = configurations.create(SettingsNames.DEPENDENCY_AGENT) {
+                asDependency()
+            }
+            dependencies.add(agentDependency.name, rootProject)
+
+            if (path == Project.PATH_SEPARATOR) {
+                configureRootProject(target, settingsExtension)
+            }
+
+            apply<KoverProjectGradlePlugin>()
+        }
+    }
+
+    private fun Project.configureRootProject(settings: Settings, settingsExtension: KoverSettingsExtensionImpl) {
+        val projectPath = path
+
+        val dependencyConfig = configurations.create(KOVER_DEPENDENCY_NAME) {
+            asDependency()
+        }
+        val rootDependencies = dependencies
+        settings.rootProject.walkSubprojects { descriptor ->
+            rootDependencies.add(KOVER_DEPENDENCY_NAME, project(descriptor.path))
+        }
+
+        val artifacts = configurations.create("koverArtifactsCollector") {
+            asConsumer()
+            attributes {
+                attribute(KoverContentAttr.ATTRIBUTE, KoverContentAttr.LOCAL_ARTIFACT)
+            }
+            extendsFrom(dependencyConfig)
+        }
+
+        val htmlTask = tasks.register<KoverHtmlReportTask>("koverHtmlReport")
+        htmlTask.configure {
+            dependsOn(artifacts)
+            this.artifacts.from(artifacts)
+            group = "verification"
+            includedProjects.convention(settingsExtension.reports.includedProjects)
+            excludedProjects.convention(settingsExtension.reports.excludedProjects)
+            excludedClasses.convention(settingsExtension.reports.excludedClasses)
+            includedClasses.convention(settingsExtension.reports.includedClasses)
+            title.convention(projectPath)
+
+            htmlDir.convention(layout.buildDirectory.dir("reports/kover/html"))
+
+            this.onlyIf {
```

Comment:
Is there a reason to use `onlyIf` for logging? Won't `doFirst/doLast` suit this purpose better?

Reply:
Because `onlyIf` always executed, even when the task does not start and the result is taken from the cache (`FROM-CACHE` outcome)

Reply:
It's worth plunking a comment describing it then